javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/indy/AdviceTransformer.java

```diff
@@ -0,0 +1,695 @@
+                  new GeneratorAdapter(mv, source.access, source.name, source.desc);
+              Type[] argumentTypes = ga.getArgumentTypes();
+              for (int i = 0; i < writableArguments.size(); i++) {
+                OutputArgument argument = writableArguments.get(i);
+                ga.dup();
+                ga.push(returnIndex + i);
+                ga.loadArg(argument.adviceIndex);
+                ga.box(argumentTypes[argument.adviceIndex]);
+                ga.arrayStore(OBJECT_TYPE);
+              }
+            }
+            super.visitInsn(opcode);
+          }
+        };
+    result = makeReadOnly(Advice.Argument.class, result);
+    return result;
+  }
+
+  /**
+   * Transform arguments annotated with {@code @Advice.Return(readOnly = false)}.
+   *
+   * <pre>{@code
+   * void foo(@Advice.Return(readOnly = false) T1 foo)
+   * }</pre>
+   *
+   * <p>is transformed to
+   *
+   * <pre>{@code
+   * @Advice.AssignReturned.ToReturned(index = 0, typing = DYNAMIC)
+   * Object[] foo(@Advice.Return(readOnly = true) T foo) {
+   *   ...
+   *   return new Object[] { foo };
+   * }
+   * }</pre>
+   */
+  @SuppressWarnings("UnusedVariable")
+  private static MethodVisitor instrumentWritableReturn(
+      MethodVisitor target, MethodNode source, OutputArgument writableReturn, int returnIndex) {
+    MethodVisitor result =
+        new MethodVisitor(Opcodes.ASM9, target) {
+          @Override
+          public void visitCode() {
+            AnnotationVisitor av =
+                visitAnnotation(Type.getDescriptor(Advice.AssignReturned.ToReturned.class), true);
+            av.visit("index", returnIndex);
+            av.visitEnum("typing", Type.getDescriptor(Assigner.Typing.class), "DYNAMIC");
+            av.visitEnd();
+            super.visitCode();
+          }
+
+          @Override
+          public void visitInsn(int opcode) {
+            if (Opcodes.ARETURN == opcode) {
+              // expecting object array on stack
+              GeneratorAdapter ga =
+                  new GeneratorAdapter(mv, source.access, source.name, source.desc);
+              Type[] argumentTypes = ga.getArgumentTypes();
+              ga.dup();
+              ga.push(returnIndex);
+              ga.loadArg(writableReturn.adviceIndex);
+              ga.box(argumentTypes[writableReturn.adviceIndex]);
+              ga.arrayStore(OBJECT_TYPE);
+            }
+            super.visitInsn(opcode);
+          }
+        };
+    result = makeReadOnly(Advice.Return.class, result);
+    return result;
+  }
+
+  /**
+   * Transform arguments annotated with {@code @Advice.Local}.
+   *
+   * <pre>{@code
+   * void foo(@Advice.Local("foo") T1 foo, @Advice.Local("bar") T1 bar)
+   * }</pre>
+   *
+   * <p>for enter advice is transformed to
+   *
+   * <pre>{@code
+   * Object[] foo(@Advice.Unused Object foo, @Advice.Unused Object bar) {
+   *   ...
+   *   Map result = new HashMap();
+   *   result.put("foo", foo);
+   *   result.put("bar", bar);
```

Comment:
Instead of referring to locals by their names, could we just put them into the array directly and refer by indexes in `OnExit`? And just keep a `<String, Integer>` mapping here, in the code generation, if we need that.

Reply:
We could and it definitely would make sense if we intended to keep this code, but as this is a temporary hack I think it isn't really important. Currently instrumenting methods doesn't have shared state, this would require having it. I'll think about it.

Reply:
Right, kinda makes sense if it's a temporary thing. Feel free to ignore it then 👍